tools/module_diag_hailcast.F90

```diff
@@ -0,0 +1,1870 @@
+!Hailcase diagnostic driver
+!John Henderson AER 20190425
+!time management handling from WRF is preserved but commented out
+
+! Yunheng Wang NSSL/CIWRO 20220310
+! Remodeled as recommended for PR #164.
+!
+
+MODULE module_diag_hailcast
+    USE time_manager_mod, ONLY: time_type, get_time
+    use mpp_mod,          only: mpp_error, FATAL, input_nml_file, stdlog, mpp_pe, mpp_root_pe
+    use fms_mod,          only: check_nml_error
+    !use fv_mp_mod,        only: is_master
+
+    use constants_mod,    only: grav, rdgas
+
+    implicit none
+
+    logical :: do_hailcast = .false. !This controls whether hailcast is used
+
+    INTEGER :: id_hailcast_dhail
+    INTEGER :: id_hailcast_dhail1, id_hailcast_dhail2, id_hailcast_dhail3, &
+               id_hailcast_dhail4, id_hailcast_dhail5, id_hailcast_wdur,   &
+               id_hailcast_wup_mask
+    INTEGER :: id_hailcast_diam_mean, id_hailcast_diam_std
+    REAL, ALLOCATABLE :: hailcast_dhail1(:,:), hailcast_dhail2(:,:),    &
+                         hailcast_dhail3(:,:), hailcast_dhail4(:,:),    &
+                         hailcast_dhail5(:,:) !hailstone diameters (mm)
+    REAL, ALLOCATABLE :: hailcast_dhail1_max(:,:), hailcast_dhail2_max(:,:), &
+                         hailcast_dhail3_max(:,:), hailcast_dhail4_max(:,:), &
+                         hailcast_dhail5_max(:,:) !hailstone diameters (mm)
+    REAL, ALLOCATABLE :: hailcast_diam_mean(:,:), hailcast_diam_std(:,:) !mean and standard deviation of five hailstones (mm)
+    REAL, ALLOCATABLE :: hailcast_wdur(:,:), hailcast_wup_mask(:,:) !persistent arrays for updraft duration (s) and mask
+    real, allocatable :: hailcast_dhail_max(:,:)
+
+    integer :: kstt_hc, kend_hc
+    integer :: kstt_hc1,kstt_hc2,kstt_hc3,kstt_hc4,kstt_hc5
+    integer :: kend_hc1,kend_hc2,kend_hc3,kend_hc4,kend_hc5
+    integer :: kstt_hcd, kend_hcd, kstt_hcm, kend_hcm
+
+    real    :: time_step
+    integer :: kdtt = 0
+
+    PRIVATE :: INTERPP, INTERP, TERMINL, VAPORCLOSE, MASSAGR, HEATBUD, BREAKUP, MELT
+    PRIVATE :: hailstone_driver, hailcast_diagnostic_driver
+    PRIVATE :: time_step, kdtt
+CONTAINS
+
+    SUBROUTINE hailcast_init(file_name, axes, Time, isco,ieco,jsco,jeco,&
+            isdo,iedo,jsdo,jedo,nlevs, missing_value, istatus)
+
+        use diag_manager_mod,   only: register_diag_field
+
+        CHARACTER(LEN=64), INTENT(IN)    :: file_name
+        integer,           intent(in)    :: axes(4)
+        type(time_type),   intent(in)    :: Time
+        INTEGER,           INTENT(IN)    :: isdo,iedo,jsdo,jedo
+        INTEGER,           INTENT(IN)    :: isco,ieco,jsco,jeco
+        INTEGER,           INTENT(INOUT) :: nlevs
+        real,              INTENT(IN)    :: missing_value
+        INTEGER,           INTENT(OUT)   :: istatus
+
+        INTEGER :: i, j
+
+        namelist /fv_diagnostics_nml/ do_hailcast
```

Comment:
Although fv_diagnostics_nml is not currently being used by the FV3 fv_diagnostics.F90 module, it should be reserved for that use.  Please rename this to a more appropriate name for the hailcast module.